release-common-lib/src/main/java/org/reactome/util/compare/DBObjectComparer.java

```diff
@@ -147,225 +173,428 @@ public static int compareInstances(
+	 * @param instance Instance to query for the values of one of its attributes
+	 * @param attribute Attribute to query for its values from the instance
+	 * @return List of generic Objects containing the values of the specific instance and attribute
+	 */
+	private static List<Object> getValues(GKInstance instance, SchemaAttribute attribute)
+	{
+		List<Object> values = getCachedValuesFromInstanceAttributeToValuesMap(instance, attribute);
+
+		if (values.isEmpty())
+		{
+			values = getRegularAttributes(instance).contains(attribute) ?
+				constructGetValuesFunctionForRegularAttributes(instance).apply(attribute) :
+				constructGetValuesFunctionForReferrerAttributes(instance).apply(attribute);
+
+			// Make sure the lists are sorted so that you are always comparing objects in the same
+			// sequence: I don't think the database adaptor applies any explicit order to Instances
+			// that don't have a rank/order attribute.
+			InstanceUtilities.sortInstances(values);
+
+			setCachedValuesFromInstanceAttributeToValuesMap(instance, attribute, values);
+		}
+		return values;
+	}
+
+	/**
+	 * Retrieves cached values for a specific attribute in a specific instance.  An empty list is returned if nothing
+	 * has yet been cached.
+	 * @param instance Instance to query for the cached values of one of its attributes
+	 * @param attribute Attribute to query for its cached values in the instance
+	 * @return List of generic Objects containing the values of the specific instance and attribute that are cached.
+	 * An empty list is returned if nothing has yet been cached.
+	 */
+	private static List<Object> getCachedValuesFromInstanceAttributeToValuesMap(
+		GKInstance instance, SchemaAttribute attribute
+	)
+	{
+		return instanceAttributeToValuesMap
+			.computeIfAbsent(instance, instanceKey -> new HashMap<>())
+			.computeIfAbsent(attribute, attributeKey -> new ArrayList<>());
+	}
+
+	/**
+	 * Caches the values passed, associating them to the specific instance and attribute passed
+	 * @param instance Instance associated with the values being cached
+	 * @param attribute Attribute associated with the values being cached
+	 * @param values Values to cache
+	 */
+	private static void setCachedValuesFromInstanceAttributeToValuesMap(
+		GKInstance instance, SchemaAttribute attribute, List<Object> values
+	)
+	{
+		instanceAttributeToValuesMap
+			.computeIfAbsent(instance, instanceKey -> new HashMap<>())
+			.put(attribute, values);
+	}
+
+	/**
+	 * Returns a function for the passed instance to retrieve values for the "regular" (i.e. not referrer) attribute
+	 * used as input to the function.
+	 * @param instance Instance the function will use to retrieve attribute values
+	 * @return Function which accepts a SchemaAttribute and returns values for the instance passed during its
+	 * construction
+	 */
+	private static Function<SchemaAttribute, List<Object>> constructGetValuesFunctionForRegularAttributes(
+		GKInstance instance
+	)
+	{
+		return (attribute) ->
```

Comment:
Why is there a lambda here? It looks like this would work fine without it. This seems like an extra layer of complexity for no clear reason, but maybe I'm missing something?

Also, same question as above but for `constructGetValuesFunctionForReferrerAttributes`...

Reply:
Originally, the logic was captured in a function and applied later in the code.  You're right, though, that now I may be able to run the code directly instead of creating and then calling the function.

Reply:
Done!